// terriajs/dist/ReactViews/Workbench/Controls/DisplayAsPercentSection.js
"use strict";
import classNames from "classnames";
import React from "react";
import createReactClass from "create-react-class";
import PropTypes from "prop-types";
import { withTranslation } from "react-i18next";
import Icon from "../../../Styled/Icon";
import Styles from "./display-as-percent.scss";
const DisplayAsPercentSection = createReactClass({
    displayName: "DisplayAsPercentSection",
    propTypes: {
        item: PropTypes.object,
        t: PropTypes.func.isRequired
    },
    togglePercentage() {
        this.props.item.displayPercent = !this.props.item.displayPercent;
    },
    render() {
        if (!this.props.item.canDisplayPercent) {
            return null;
        }
        const { t } = this.props;
        return (React.createElement("label", { className: Styles.main },
            React.createElement("button", { type: "button", onClick: this.togglePercentage, className: classNames(Styles.btn, {
                    [Styles.btnActive]: this.props.item.displayPercent,
                    [Styles.btnInactive]: !this.props.item.displayPercent
                }) },
                this.props.item.displayPercent ? (React.createElement(Icon, { glyph: Icon.GLYPHS.checkboxOn })) : (React.createElement(Icon, { glyph: Icon.GLYPHS.checkboxOff })),
                React.createElement("span", null, t("workbench.displayPercent")))));
    }
});
module.exports = withTranslation()(DisplayAsPercentSection);
